daffodil-japi/src/main/scala/org/apache/daffodil/japi/Daffodil.scala

```diff
@@ -866,6 +867,12 @@ class ParseResult private[japi] (pr: SParseResult, deprecatedOutput: Maybe[JDOMI
    * @return true if any validation errors occurred, false otherwise.
    */
   def isValidationError(): Boolean = pr.isValidationError
+
+  /**
+   * Optional result of the infoset validation.
+   * @return optional ValidationResult
+   */
+  def validationResult(): Option[ValidationResult] = pr.validationResult()
```

Comment:
Ah, I somehow missed the sapi/japi changes. Two concerns here though:

1. ``Option`` is a Scala thing that isn't familiar to Java users so should be avoided in the JAPI. This API should follow Java conventions, which I guess means return null if there is no ValidationResult?

2. The ValidationResult that is returned is in the org.apache.daffodil.api package. But the Java/Scala doc that we generate only comes from the org.apache.daffodil.sapi/japi packages. This means no documentation will be generated for the ValidationResult, so users wont' really know how to use it. The way we have been maintaining the Java and Scala API is to create wrappers in sapi/japi that are essentially proxies for classes outside of those packages. So there would be a japi.ValidationResult and sapi.ValidationResult which wrap the api.ValidationResult, but have java/scaladoc that the user will see. It's kindof terrible, and we have tickets to fix this so there's only a single API and not all these proxy classes, but it's the standard for what we have right now. It's does also have a nice benefit in that it makes it very clear what is public API and what isn't (i.e. if it's in the org.apache.daffodil.sapi/japi then it's public), and clearly defines when we need to be concerned with backwards compatibility.

Reply:
So looking at what is involved here, and thinking about the comment above...

> Is the intention to keep this a sort of internal/experimental API until all the new validation stuff is run through its paces?

The only immediate use of this raw capability is for the CLI.  Is there any value in publishing this API to the SAPI and JAPI now?  Or would it be just as well to let it internal until there is demand.

I lean toward removing the additions I made to SAPI/JAPI and only exposing it through the CLI for now.

Reply:
> Personally, I would like to see more use-cases of this rawValidation data before we make it part of the public API.

Concur.

> like maybe we want alternative syntax like --validate=foo.sch:output=details.svrl

Ive went down this road, there was an implementation in the original validator refactorings that support syntax similar to this.

Question is/was, does it scale?  I say not beyond a low number of shortish parameters.

Still valuable to input arbitrary configuration to validators, so there is [support for this by passing a config file](https://github.com/apache/daffodil/blob/master/daffodil-cli/src/main/scala/org/apache/daffodil/Main.scala#L233) that was added in the validator work.

But Id say that doesnt help here, because managing the output files down in the validator creates issues with reusing the validators, which is important part of that design.  So the data has to get up out of the validators to be written.

> Thoughts?

The raw stream wasnt the first choice, but its not so unappealing if you squint.  Conceptually there are two streams of data that come out of validators (1) parsed structured data by way of validation result (2) optional raw source data that the structured data was parsed from.

Reply:
> Question is/was, does it scale? I say not beyond a low number of shortish parameters.

Fair, but I don't think the current solution scales all that well either--it only supports a single option and a single output stream.

As an alternative to what I suggested earlier, maybe we do something where the .conf file (not sure what the syntax is for specifying that, can' find any test) can/should be used for configurations that are static (e.g. an encoding option, enable/disable features), but additional options can be provided to override those, or when more dynamic options are needed like in this case. We could also use an alternative syntax that is maybe less ugly, for example:
```
--validate foo.sh --validate-option config=foo.conf --validate-option output=details.svrl
```
We have very similar syntax for defining external variables, e.g. ``-Dvariable1=value1 -Dvariable2=value2``

So ``--validate`` can be provided once to specify the mode/validator plugin, and ``--validate-option`` is a key-value pair that can be specified multiple times for different options. One option could provide a config file to configure the validator, other options could override those when needed or to provide more dynamic options.

An added benefit (depending on how you view it?) is that now validators are responsible for supporting/parsing the config option, and so they can use whatever configuration format makes sense for them. Maybe some validators already have a well defined and popular config file format that is different than what we currently support. If these are truely pluggable, we might not want to force a specific format.

> Conceptually there are two streams of data that come out of validators

While I agree that's probably the case, I'm not sure a byte array is always the most natural way to get the second thing that could come out of a validator. I'd argue when using the API, we never want a byte array--we usually want specific underlying data structures with casting probably being the best way to get to them. And when not using the API, then I'd rather assume the validator knows best what information should be output, where, and what the most efficient way to do it, and using some config stuff to tweak it.

All that said, if you're not sold, I won't stand in the way of this change. This definitey provides a useful feature that we do need, and it is well done. And if we do eventually decide we need something different is this is run through its paces, we can always go through the normal deprecation process. For example, it would be pretty trivial to to change ``--validate-output=foo`` to ``--validate-option output=foo`` behind the scenes.

+1 :+1:

Reply:
Thats all sane.  Ill take a look at refactoring towards what you describe.

> Fair, but I don't think the current solution scales all that well either--it only supports a single option and a single output stream.

You just arent squinting hard enough.  :grin:

Reply:
Have wrestled with the refactors here, they grow the size of this PR pretty quick.  I am thinking this to get this functionality in but avoid adding things we dont need.

1. Drop the raw stream, the validator can write the output based on config passed into it
2. Revert all changes to CLI, we already have support for the conf file being passed.  Ill write a test for that and it can be used to configure the validator to output svrl for now.  The refactor to multiple validator args can be separate PR.
3. Keep the other refactors in Schematron and Parse Result

Ill follow up with a ticket something like "Refactor validator CLI argument passing" with the description from above.